# Download NYC Mayoral Election Data

We will use election results form the 2025 NYC Mayoral Elections at the ED/AD (Election District/Assembly District) level:
- Primary Election Results ([NYTimes](https://www.nytimes.com/interactive/2025/06/24/us/elections/nyc-mayor-primary-results-precinct-map.html)) ([CUNY](https://www.electionatlas.nyc/maps.html#!NYCResults2025primaryinteractive))
- General Election Results ([NYTimes](https://www.nytimes.com/interactive/2025/11/04/us/elections/nyc-mayor-results-precinct-map.html)) ([CUNY](https://www.electionatlas.nyc/maps.html#!NYCResults2025generalinteractive))

Election District (ED) is the smallest unit for which election results are reported.
Assembly District (AD) is a larger unit that contains multiple EDs.

I've provided you election data in the file `NYC Mayoral Election 2025.xlsx`. 
But since the Demographic data is only available at the AD level, we will need load it and aggregate the election data up to the AD level as well.

## Setup

You can ignore this secton.

This setup allows you to use *Python* and *R* in the same notebook.

To set up a similar notebook, see quickstart instructions here:

https://github.com/dmil/jupyter-quickstart

Notes on the benefits of this setup are in that repo as well.

In [1]:
%load_ext rpy2.ipython
%load_ext autoreload
%autoreload 2

%matplotlib inline  
from matplotlib import rcParams
rcParams['figure.figsize'] = (16, 100)

import warnings
from rpy2.rinterface import RRuntimeWarning
warnings.filterwarnings("ignore") # Ignore all warnings
# warnings.filterwarnings("ignore", category=RRuntimeWarning) # Show some warnings

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display, HTML

In [2]:
%%javascript
// Disable auto-scrolling
IPython.OutputArea.prototype._should_scroll = function(lines) {
    return false;
}

<IPython.core.display.Javascript object>

This is a Python notebook, but below is an R cell. The `%%R` at the top of the cell indicates that the code in this cell will be R code.

In [3]:
%%R

# My commonly used R imports
require('tidyverse')

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.2.1
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


Loading required package: tidyverse


# Load Data

### Load General Election Data

In [4]:
df_general = pd.read_excel("NYC Mayoral Election 2025.xlsx", sheet_name="General (Unofficial)")
df_general

aded25  gen25zm  gen25ac  gen25cs  gen25ea  gen25othr  gen25tot    win25
0      23001       78      662      346        1          7      1094  gen25ac
1      23002       75      627      369        1          6      1078  gen25ac
2      23003       33      184      157        0          2       376  gen25ac
3      23004      102      757      226        3          4      1092  gen25ac
4      23005       82      731      274        1          3      1091  gen25ac
...      ...      ...      ...      ...      ...        ...       ...      ...
4333   87051      136      148       11        1          3       299  gen25ac
4334   87052      105       96        5        1          1       208  gen25zm
4335   87053        0        0        0        0          0         0      Tie
4336   87054        0        0        0        0          0         0      Tie
4337   87055        0        0        0        0          0         0      Tie

[4338 rows x 8 columns]

- `aded25` is the 2025 assembly district code + election distinct code. (23001) means assembly district 23, election district 001.
- `gen_25_zm` is the number of votes **Zohran Mamdani** received in the general election.
- `gen25ac` is the number of votes **Andrew Cuomo** received in the general election.
- `gen25cs` is the number of votes **Curtis Sliwa** received in the general election.
- `gen25ea` is the number of votes **Eric Adams** received in the general election.

### Load Primary Election Data

In [5]:
df_primary = pd.read_excel("NYC Mayoral Election 2025.xlsx", sheet_name="Primary")
df_primary

districtid  ad  ed  lowtwo1  adamsa1  cuomo1  lander1  ramos1  blake1  \
0          23055  23  55        0        0      19        0       1       0   
1          23056  23  56        0        1      32        1       0       2   
2          23057  23  57        1        1      34        3       0       0   
3          23058  23  58        1        1      23        0       0       0   
4          23059  23  59        0        5      32        1       0       0   
...          ...  ..  ..      ...      ...     ...      ...     ...     ...   
4031       87048  87  48        0       10     111        3       1       3   
4032       87049  87  49        0       30     208        4       4       4   
4033       87050  87  50        0        8      69        0       2       0   
4034       87051  87  51        0       12      70        3       1       2   
4035       87052  87  52        0        4      64        2       0       1   

      stringer1  ...  lander9  ramos9  blake9  stringer9  tilson9  myrie9  \
0             1  ...        0       0       0          0        0       0   
1             1  ...        0       0       0          0        0       0   
2             1  ...        0       0       0          0        0       0   
3             3  ...        0       0       0          0        0       0   
4             2  ...        0       0       0          0        0       0   
...         ...  ...      ...     ...     ...        ...      ...     ...   
4031          3  ...        0       0       0          0        0       0   
4032          2  ...        0       0       0          0        0       0   
4033          2  ...        0       0       0          0        0       0   
4034          0  ...        0       0       0          0        0       0   
4035          0  ...        0       0       0          0        0       0   

      mamdani9  usedvotes9  inactive9  total9  
0           43          62          2      64  
1           33          66          0      66  
2           73         107          3     110  
3           34          58          3      61  
4           12          44          6      50  
...        ...         ...        ...     ...  
4031        89         206          9     215  
4032        70         293         16     309  
4033        36         107          6     113  
4034        31         110          9     119  
4035        25          91          5      96  

[4036 rows x 121 columns]

New York City uses ranked choice voting (aka instant runoff voting) in its primaries. This means that voters rank their choices. Vote tabulation is done in rounds (so in the first round, the person with the feweset votes is eliminated, and their votes are redistributed to the second choice on those ballots, and so on). The dat acontains all 9 rounds of tabulation. So `mamdani1` is the number of votes Zohran Mamdani received in round 1, `mamdani9` is the number of votes he received in the final round (round 9).

# Clean & Combine Data

Since Demographic data are only available at the assembly district level, we will roll up the rest of the data to that level also.

### Clean Primary Data

In [6]:
# Aggregate Primary Data by Assembly District
df_primary_by_ad = df_primary.pivot_table(index='ad', values=['mamdani9','cuomo9', 'total9'], aggfunc='sum')

# rename columns
df_primary_by_ad.rename(columns={
    'mamdani9': 'mamdani_primary',
    'cuomo9': 'cuomo_primary',
    'total9': 'total_primary'
    }, inplace=True)
df_primary_by_ad 

cuomo_primary  mamdani_primary  total_primary
ad                                               
23           6688             3871          11130
24           5549             6233          12200
25           4329             3928           8648
26           7190             5296          13244
27           6314             3783          10555
..            ...              ...            ...
83           8234             3280          12099
84           4812             3872           9088
85           5121             2990           8551
86           4446             2610           7369
87           5294             5087          10770

[65 rows x 3 columns]

### Clean General Data

In [7]:
# get the first two characters of df_general['aded25']
df_general['ad'] = df_general['aded25'].astype(str).str[:2].astype(int)
df_general_by_ad = df_general.pivot_table(index='ad', values=['gen25zm','gen25ac','gen25cs','gen25ea','gen25othr','gen25tot'], aggfunc='sum')

df_general_by_ad.rename(columns={
    'gen25zm': 'mamdani_general',
    'gen25ac': 'cuomo_general',
    'gen25cs': 'sliwa_general',
    'gen25ea': 'adams_general',
    'gen25othr': 'other_general',
    'gen25tot': 'total_general'
}, inplace=True)

df_general_by_ad = df_general_by_ad[['mamdani_general', 'cuomo_general', 'sliwa_general', 'adams_general', 'other_general', 'total_general']]

df_general_by_ad

mamdani_general  cuomo_general  sliwa_general  adams_general  \
ad                                                                 
23             9072          19096           7086             82   
24            13825          11985           2256             66   
25             8420          13485           2485             59   
26             9978          20901           5205             70   
27             7624          16738           3495             71   
..              ...            ...            ...            ...   
83             9928           7198            616            102   
84            10111           6701            810             84   
85             8957           7222            923             86   
86             7322           5951            664             81   
87            12676           7487           1090             90   

    other_general  total_general  
ad                                
23            152          35488  
24            113          28245  
25            154          24603  
26            196          36350  
27            136          28064  
..            ...            ...  
83             94          17938  
84            121          17827  
85            111          17299  
86             97          14115  
87            145          21488  

[65 rows x 6 columns]

### Combine Data

In [8]:
df = df_primary_by_ad\
    .merge(df_general_by_ad, left_index=True, right_index=True) \

df

cuomo_primary  mamdani_primary  total_primary  mamdani_general  \
ad                                                                   
23           6688             3871          11130             9072   
24           5549             6233          12200            13825   
25           4329             3928           8648             8420   
26           7190             5296          13244             9978   
27           6314             3783          10555             7624   
..            ...              ...            ...              ...   
83           8234             3280          12099             9928   
84           4812             3872           9088            10111   
85           5121             2990           8551             8957   
86           4446             2610           7369             7322   
87           5294             5087          10770            12676   

    cuomo_general  sliwa_general  adams_general  other_general  total_general  
ad                                                                             
23          19096           7086             82            152          35488  
24          11985           2256             66            113          28245  
25          13485           2485             59            154          24603  
26          20901           5205             70            196          36350  
27          16738           3495             71            136          28064  
..            ...            ...            ...            ...            ...  
83           7198            616            102             94          17938  
84           6701            810             84            121          17827  
85           7222            923             86            111          17299  
86           5951            664             81             97          14115  
87           7487           1090             90            145          21488  

[65 rows x 9 columns]

In [9]:
# write to csv
df.to_csv('data/election_data.csv')